tests/surrogate_test.py

```diff
@@ -0,0 +1,206 @@
+    assert numpy.allclose(f._mapdomain([-3, 7],
+                                       domain, new_domain), [0.25, 0.75])
+
+
+def test_map_function_2d():
+    """Test if points are transfromed properly."""
+    grid = SmolyakGridGenerator(ChebyshevFirstKind.make_nested_set(1))
+    domain = ((-10, 10), (0, 2))
+    new_domain = ((0, 1), (-1, 1))
+    f = Surrogate(domain, grid)
+    assert numpy.allclose(f._mapdomain((-10, 0), domain, new_domain), [0, -1])
+    assert numpy.allclose(f._mapdomain([(0, 1), (10, 2)],
+                                       domain, new_domain), [[0.5, 0], [1, 1]])
+
+
+def test_basis_matrix():
+    """Test if a correct basis matrix is generated."""
+    grid_generator = SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(1))
+    surrogate_object = Surrogate([(-7, 3), (-18, 9)], grid_generator)
+    expected_basis_matrix = numpy.array(
+        [[1, 0, -1, 0, -1], [1, -1, 1, 0, -1], [1, 1, 1, 0, -1],
+         [1, 0, -1, -1, 1], [1, 0, -1, 1, 1]])
+    assert numpy.allclose(expected_basis_matrix,
+                          surrogate_object._make_basis_matrix())
+
+
+def test_surrogate_0():
+    """Test if surrogate generates exact results for test fucntion 0."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1), (-1, 1)],
+                                 SmolyakGridGenerator(
+                                     ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_0, 'lu')
+    test_point = (0.649, 0, -0.9)
+    test_point_ = (-0.885, 1, 0.275)
+    assert numpy.allclose(surrogate_object(test_point),
+                          function_0([0.649, 0, -0.9]))
+    assert numpy.allclose(surrogate_object(test_point_),
+                          function_0([-0.885, 1, 0.275]))
+
+
+def test_surrogate_1():
+    """Test if surrogate generates exact results for test for Chebyshevs."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1)], SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_1, 'lstsq')
+    test_point = (0.649, -0.9)
+    test_point_ = (-0.885, 1)
+    assert numpy.allclose(surrogate_object(test_point),
+                          function_1([0.649, -0.9]))
+    assert numpy.allclose(surrogate_object(test_point_),
+                          function_1([-0.885, 1]))
+
+
+def test_surrogate_0_shifted():
+    """Test if surrogate generates exact results for a shifted function 0."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1), (-1, 1)],
+                                 SmolyakGridGenerator(
+                                     ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object_ = Surrogate([(0, 2), (-2, 0), (-1, 1)],
+                                  SmolyakGridGenerator(
+                                      ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_0, 'lu')
+    surrogate_object_.train(function_0_shifted, 'lu')
```

Comment:
I didn't notice before, but this is a case where it would be helpful to have saved the training data:
```suggestion
    surrogate_object.train(function_0, 'lu')
    surrogate_object_.train_from_data(surrogate_object.data, 'lu')
```
You should go add this is as a property of the surrogate, and set its value in `train_from_data`.

Reply:
Yes it can. You called `surrogate_object.train` in line 161. If you store its `data` in a property as suggested, it will be accessible when you call `surrogate_object_.train_from_data` (which is a different surrogate) on line 162.